src/java/org/apache/cassandra/service/consensus/migration/TableMigrationState.java

```diff
@@ -0,0 +1,345 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service.consensus.migration;
+
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.tcm.Epoch;
+import org.apache.cassandra.tcm.serialization.MetadataSerializer;
+import org.apache.cassandra.tcm.serialization.Version;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.apache.cassandra.db.TypeSizes.sizeof;
+import static org.apache.cassandra.dht.Range.intersectionOfNormalizedRanges;
+import static org.apache.cassandra.dht.Range.normalize;
+import static org.apache.cassandra.dht.Range.subtract;
+import static org.apache.cassandra.dht.Range.subtractNormalizedRanges;
+import static org.apache.cassandra.utils.CollectionSerializers.deserializeMap;
+import static org.apache.cassandra.utils.CollectionSerializers.deserializeSet;
+import static org.apache.cassandra.utils.CollectionSerializers.newHashMap;
+import static org.apache.cassandra.utils.CollectionSerializers.serializeCollection;
+import static org.apache.cassandra.utils.CollectionSerializers.serializeMap;
+import static org.apache.cassandra.utils.CollectionSerializers.serializedCollectionSize;
+import static org.apache.cassandra.utils.CollectionSerializers.serializedMapSize;
+
+// TODO Move this into the schema for the table once this is based off of TrM
+public class TableMigrationState
+{
+    private static final Logger logger = LoggerFactory.getLogger(TableMigrationState.class);
+
+    @Nonnull
+    public final String keyspaceName;
+
+    @Nonnull
+    public final String tableName;
+
+    @Nonnull
+    public final TableId tableId;
+
+    @Nonnull
+    public final ConsensusMigrationTarget targetProtocol;
```

Comment:
I think I'd prefer to keep this as is. Since epochs can change during the processing of a query, I think it will be less work to keep this in sync than it will be to ensure you're looking at the migration state and table metadata from the same epoch. I added a sanity check to the ClusterMetadata build method to check that they're in sync